glasbey.py

```diff
@@ -38,118 +45,227 @@ def finish(self):
+                )
+                pbar.update(i)
+                i += 1
+        pbar.finish()
+        return colors
+
+    def extend_palette(self, size):
+        """
+        Add new, distinct colors to self.palette.
+        """
+        assert size > len(self.palette)  # sanity check
+
+        # Initialize distances array
+        num_colors = self.colors.shape[0]
+        distances = np.ones(shape=(num_colors, 1)) * 1000
+
+        # A function to recompute minimum distances from palette to all colors
+        def update_distances(colors, color):
+            d = np.linalg.norm((colors - color), axis=1)
+            np.minimum(distances, d.reshape(distances.shape), distances)
+
+        # Build progress bar
+        widgets = ["Generating palette: ", Percentage(), " ", Bar(), " ", ETA()]
+        pbar = ProgressBar(widgets=widgets, maxval=size).start()
+        # Update distances for the colors that are already in the palette
+        for i in range(len(self.palette) - 1):
+            update_distances(self.colors, self.palette[i])
+            pbar.update(i)
+        # Iteratively build palette
+        while len(self.palette) < size:
+            update_distances(self.colors, self.palette[-1])
+            self.palette.append(self.colors[np.argmax(distances), :])
+            pbar.update(len(self.palette))
+        pbar.finish()
+
+        assert self.palette_invariant(), "Internal error during extend_palette: self.palette is poorly formatted."
+
+        if self.overwrite_base_palette:
+            self.save_palette(palette=self.palette, path=self.base_palette, format="byte", overwrite=True)
+
+    def load_palette(self, path):
+        """
+        Expected format: sRGB255
+        """
+        assert os.path.isfile(path)
+        palette = list()
+        with open(path, 'r') as file:
+            for line in file:
+                rgb = [int(c) for c in line.strip().split(",")]
+                palette.append((rgb[0] * 256 + rgb[1]) * 256 + rgb[2])
+        return palette
+
+    @staticmethod
+    def save_palette(palette, path: str, format: str = "byte", overwrite: bool = False):
+        """
+        Output format examples (white):
+            * byte:  255,255,255  (sRGB255)
+            * float: 1.000000,1.000000,1.000000
+        """
+        if not overwrite:
+            assert not os.path.isfile(path)
+
+        with open(path, 'w') as file:
+            if format.lower() == "byte":
+                for color in palette:
+                    rgb255 = tuple(int(round(k * 255)) for k in color)
+                    file.write("{},{},{}\n".format(*rgb255))
+            elif format.lower() == "float":
+                for color in palette:
+                    file.write("{:.6f},{:.6f},{:.6f}\n".format(*(abs(k) for k in color)))
+            else:
+                raise ValueError("Format doesn't match. Choose between 'byte' and 'float'")
+
+    def palette_invariant(self):
```

Comment:
Are you sure "invariant" is what you meant? For me rather looks like the function is answering question "is palette valid?"

Reply:
Renamed the functions to...
check_validity_rbg_palette()
check_validity_internal_palette()